File: Sorting/InsertionSort.java
// We take an element from the unsorted part and insert it in the right place in the sorted part
// Initially one element would be sorted because a single elment is always sorted in itself

// Time Complexity O(n^2)
// Space Complexity O(1)
// In-place, Stable, Adaptive, Online

public static void insertionSort(int [] arr){
    for(int i = 1; i < arr.length; i++){
        int current = arr[i];
        int j = i - 1;
        while(arr[j] > current && j >= 0){
            arr[j+1] = arr[j];
            j--;
        }
        arr[j+1] = current;
    }
}

File: Sorting/InsertionSort.js
// We take an element from the unsorted part and insert it in the right place in the sorted part
// Initially one element would be sorted because a single elment is always sorted in itself

// Complexity          Best                        Average                    Worst

// Time                O(n)                        O(n^2)                     O(n^2)

// Space               O(1)                         O(1)                       O(1)

// Stable, Adaptive, Online

function insertionSort(arr) {
    for (let i = 1; i < arr.length; i++) {
        let j = i;
        while (arr[j - 1] > arr[j] && j >= 0) {
            [arr[j - 1], arr[j]] = [arr[j], arr[j - 1]];
            j--;
        }
    }
    return arr;
}
